### ISSUE TITLE
Tune `open_zarr()` chunks parameter, providing good defaults.

### ISSUE THREAD

**Issue by alxmrs** [m2lines/Samudra#106]
I'm working on a simple PR that will make the `chunks={}` argument configurable. From meeting with @jder yesterday, it seems like `chunks=dict(time=700)` should lead to some improvement; how much is a matter of measuring properly. To that end, in addition to parameterizing this, I'm messing a bit with the trainer's logging system.

In the future, this could be a hparam that we perform sweeps on (which also is dependent on the hw the trainer runs on).

**Comment by jder** [comment]
Starting with some context: profiling the om4 training run on a `a2-highgpu-1g` with 1 A100 shows that each training step spends ~2/3 of its time waiting for data loaders (using 4 data loaders). https://openathena.slack.com/archives/C08CYM42DT3/p1741385964346899

As a first cut, we tried comparing:
* [Baseline](https://github.com/suryadheeshjith/Ocean_Emulator/tree/u/jder/experiments/profile-2025-03-11) with a single in-thread data loader, training on a single `a2-highgpu-1g` instance on the om4 data, one epoch of 16 steps. 
* Same as baseline, but in the open_zarr calls, passing `chunks=dict(time="100mb")` to try and get it to load more data at once.
* Same as baseline but pointing at underling zarr files rechunked to have chunks with t=700. 

Result was
* Baseline: train time 131.89s
* t100mb on read: train time 129.23s
* t=700 rechunked: train time 130.30
* Baseline again: train time 128.08

**Comment by alxmrs** [comment]
Is train time the right metric? I think we want to be measuring the length of the delay before the first data allocation on to the GPU (or something similar). 

BTW, can I fork off of your branch?

**Comment by jder** [comment]
We can certainly measure that but wouldn’t you expect overall train time to go down too? (Which seems more directly like the thing that matters). Feel free to fork off the branch.

**Comment by alxmrs** [comment]
I might expect that. How is train time measured? To me, it seems to not include any data loading / preparation stuff. e.g., it's not recording any of the GPU idle time.

**Comment by jder** [comment]
This is the train time -- it's wall-clock time for the epoch https://github.com/suryadheeshjith/Ocean_Emulator/blob/3eb5b3bb7d123cfc0adcd72a5264fa0d68b25c5e/src/train_3D.py#L416-L418

**Comment by alxmrs** [comment]
ah. This seems to confirm my suspicions. I don't think it's capturing the idling GPU.

**Comment by suryadheeshjith** [comment]
I’ve observed that the performance of reading a zarr dataset varies depending on whether the chunking scheme used during saving matches the chunking scheme used during reading. (I could be wrong)

For CM4, on the clusters, I saved the files with the following chunking:
chunks = {"time": 10, "lat": 180, "lon": 360}

When reading the Zarr, I use chunks={}.

**Comment by jder** [comment]
Three more experiments this morning.

I tried [updating all libraries](https://github.com/suryadheeshjith/Ocean_Emulator/tree/2692107279013565e444618c5cc7b697ac06668c) on top of the baseline above, including to zarr-3 and python 11, and I tried adding a sleep(1) to `TrainDataset.__get_item__` to verify we're measuring what I think we're measuring, and then I tried update all + the rechunking on disk with t=700 change from above:

* Update everything: train time 134.62s
* Update everything + add sleep(1): train time 148.10s
* ~Update everything + t700: train time 134.85~ EDIT: later discovered this was not actually chunked at t=700
* (Compare with Baseline from above): 131.89s

**Comment by jder** [comment]
I tried [preloading all the data](https://github.com/suryadheeshjith/Ocean_Emulator/tree/1d18661377ffd28232807f94d03698e7deb67239) into memory today. This takes about 3 seconds on my test machine at the start of the training run. 

* Preload all: train time **18.5**
* Baseline again for paranoia: train time 125.48

Seems like a big win, next step is to confirm this in a more realistic training environment.

**Comment by jder** [comment]
Ok, I tried a [more realistic baseline](https://github.com/suryadheeshjith/Ocean_Emulator/tree/d62ddf3599f9ccb397f8faf29965426d14b55f70) which I think is the full samudra model but with smaller batch sizes since I'm on a 40GB A100.

Batch size 1: 
* Baseline with full model: train time 133.69
* Preload data with full model: train time 24.57

Batch size 2:
* Baseline with full model: train time 132.75
* Preload data with full model: train time 24.47

Seems a little surprising that batch size increases didn't change total time (you would expect to get wins from more parallelism) so maybe something fishy in my setup or else another opportunity to see if we can optimize. 

(Sorry I guess this is ~offtopic for your issue now, I'll move this to a new issue+PR)

**Comment by jder** [comment]
Let's do a sweep of this based on train dataloading time once we have the new xarray-based data loader.

**Comment by alxmrs** [comment]
As @jder has discovered, increasing time dim chunks is foiled by the random shuffle over time. If chunks were to pay off, we would need to pre-shuffle the dataset. Further, we get a benefit from grouping all the level information into one IO call, but that also requires data engineering (i.e. "data compaction" -- see #120).
